### ISSUE TITLE
resource handles non-const get()

### ISSUE THREAD

**Issue by huwpascoe** [skypjack/entt#278]
feature request:

Some resources can be modified at runtime, the current implementation in handle.hpp doesn't account for this. It'd be nice to have a non-const pointer as an option.

```
const Resource & get() const ENTT_NOEXCEPT;
Resource & get() const ENTT_NOEXCEPT;
```

**Comment by Kerndog73** [comment]
https://github.com/skypjack/entt/blob/a470925305cffdf2ea41828ccc40e7e07e36d006/src/entt/resource/handle.hpp#L55

**Comment by skypjack** [comment]
What's wrong with the function pointed out by @Kerndog73 ?

**Comment by huwpascoe** [comment]
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const

**Comment by huwpascoe** [comment]
Hang on, this doesn't look the same as the version I'm using...

**Comment by huwpascoe** [comment]
I'm using version 3.0.0. Is the master branch just different?

```
#ifndef ENTT_RESOURCE_HANDLE_HPP
#define ENTT_RESOURCE_HANDLE_HPP


#include <memory>
#include <utility>
#include "../config/config.h"
#include "fwd.hpp"


namespace entt {


/**
 * @brief Shared resource handle.
 *
 * A shared resource handle is a small class that wraps a resource and keeps it
 * alive even if it's deleted from the cache. It can be either copied or
 * moved. A handle shares a reference to the same resource with all the other
 * handles constructed for the same identifier.<br/>
 * As a rule of thumb, resources should never be copied nor moved. Handles are
 * the way to go to keep references to them.
 *
 * @tparam Resource Type of resource managed by a handle.
 */
template<typename Resource>
class resource_handle {
    /*! @brief Resource handles are friends of their caches. */
    friend class resource_cache<Resource>;

    resource_handle(std::shared_ptr<Resource> res) ENTT_NOEXCEPT
        : resource{std::move(res)}
    {}

public:
    /*! @brief Default constructor. */
    resource_handle() ENTT_NOEXCEPT = default;

    /**
     * @brief Gets a reference to the managed resource.
     *
     * @warning
     * The behavior is undefined if the handle doesn't contain a resource.<br/>
     * An assertion will abort the execution at runtime in debug mode if the
     * handle is empty.
     *
     * @return A reference to the managed resource.
     */
    const Resource & get() const ENTT_NOEXCEPT {
        ENTT_ASSERT(static_cast<bool>(resource));
        return *resource;
    }

    /**
     * @brief Casts a handle and gets a reference to the managed resource.
     *
     * @warning
     * The behavior is undefined if the handle doesn't contain a resource.<br/>
     * An assertion will abort the execution at runtime in debug mode if the
     * handle is empty.
     */
    inline operator const Resource &() const ENTT_NOEXCEPT { return get(); }

    /**
     * @brief Dereferences a handle to obtain the managed resource.
     *
     * @warning
     * The behavior is undefined if the handle doesn't contain a resource.<br/>
     * An assertion will abort the execution at runtime in debug mode if the
     * handle is empty.
     *
     * @return A reference to the managed resource.
     */
    inline const Resource & operator *() const ENTT_NOEXCEPT { return get(); }

    /**
     * @brief Gets a pointer to the managed resource from a handle.
     *
     * @warning
     * The behavior is undefined if the handle doesn't contain a resource.<br/>
     * An assertion will abort the execution at runtime in debug mode if the
     * handle is empty.
     *
     * @return A pointer to the managed resource or `nullptr` if the handle
     * contains no resource at all.
     */
    inline const Resource * operator->() const ENTT_NOEXCEPT {
        ENTT_ASSERT(static_cast<bool>(resource));
        return resource.get();
    }

    /**
     * @brief Returns true if a handle contains a resource, false otherwise.
     * @return True if the handle contains a resource, false otherwise.
     */
    explicit operator bool() const { return static_cast<bool>(resource); }

private:
    std::shared_ptr<Resource> resource;
};


}


#endif // ENTT_RESOURCE_HANDLE_HPP
```

**Comment by skypjack** [comment]
> http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const

Too late, I like this pattern. :-)  
Yes, `master` contains the code for the upcoming version 3.1 and this function wasn't present in the version 3.0 actually.  
You can either use `master` or wait until v3.1 is released.

**Comment by huwpascoe** [comment]
Disregard that link, I thought the implication was that I myself had to use const_cast!

Feature already implemented 👍 I'll wait until 3.1.

**Comment by skypjack** [comment]
Consider that `master` is almost always a stable branch and you can safely use it if you want.  
I usually implement new features on other branches before to merge them on `master`. :+1:
